cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+        u = self.s ^ (self.F[p,:] & (~self.v)) ^ (self.M[p,:] & self.v)
+        
+        alpha = sum(self.G[p,:] & (~self.v) & self.s) % 2
+        beta = sum(self.M[p,:] & (~self.v) & self.s) 
+        beta += sum(self.F[p,:] & self.v & self.M[p,:])
+        beta += sum(self.F[p,:] & self.v & self.s)
+        beta %= 2
+        
+        delta = (self.gamma[p] + 2*(alpha+beta)) % 4
+        
+        self._update_sum(t,u,delta=delta,alpha=alpha)
+    
+    def _update_sum(self,t,u,delta=0,alpha=0):
+        ''' Implements the transformation (Proposition 4 in Bravyi et al)
+
+                i^alpha U_H (|t> + i^delta |u>) = omega W_C W_H |s'>
+        '''
+        if np.all(t == u):
+            self.s = t
+            self.omega *= 1/np.sqrt(2)*(-1)**alpha*(1+1j**delta)
+        else:
+            set0 = np.where((~self.v) & (t ^ u))[0]
+            set1 = np.where(self.v & (t ^ u))[0]
+            
+            # implement Vc
+            if len(set0) > 0:
+                q = set0[0]
+                for i in set0:
+                    if i != q:
+                        self._CNOT(q,i,right=True)
+                for i in set1:
+                    self._CZ(q,i,right=True)
+            elif len(set1) > 0:
+                q = set1[0]
+                for i in set1:
+                    if i != q:
+                        self._CNOT(i,q,right=True)
+                
+            e = np.zeros(self.n,dtype=bool)
+            e[q] = True
+            
+            if t[q]:
+                y = u ^ e
+                z = u
+            else:
+                y = t
+                z = t ^ e
+                
+            (omega,a,b,c) = self._H_decompose(self.v[q],y[q],z[q],delta)
+            
+            self.s = y
+            self.s[q] = c
+            self.omega *= (-1)**alpha*omega
+            
+            if a:
+                self._S(q,right=True)
+            self.v[q] ^= b ^ self.v[q]
+
+    def _H_decompose(self,v,y,z,delta):
+        """ Determines the transformation 
+
+                H^v (|y> + i^delta |z>) = omega S^a H^b |c>
+
+        where the state represents a single qubit. 
+
+        Input: v,y,z are boolean; delta is an integer (mod 4)
+        Outputs: a,b,c are boolean; omega is a complex number
+        
+        Precondition: y != z """
+        if y == z:
+            raise ValueError('|y> is equal to |z>')
+
+        if not v:
+            omega = (1j)**(delta*int(y))
+
+            delta2 = ((-1)**y * delta) % 4
+            c = bool((delta2 >> 1))
+            a = bool(delta2 & 1)
+            b = True
+        else:
+            if not (delta & 1):
+                a = False
+                b = False
+                c = bool(delta >> 1)
+                omega = (-1)**(c & y)
+            else:
+                omega = 1/np.sqrt(2)*(1+1j**delta)
+                b = True
+                a = True
+                c = not ((delta >> 1) ^ y)
+
+        return omega,a,b,c
+
+    def to_numpy(self):
```

Comment:
Document, add type annotations, give a better name (e.g. `to_state_vector`)

Reply:
Renaming to `to_state_vector` but not documenting because the name documents it.